apps/src/blockly/addons/plusMinusBlocks/text_join.js

```diff
@@ -0,0 +1,196 @@
+   * @this {GoogleBlockly.Block}
+   */
+  domToMutation: function (xmlElement) {
+    const targetCount = parseInt(xmlElement.getAttribute('items'), 10);
+    this.updateShape_(targetCount);
+  },
+
+  /**
+   * Returns the state of this block as a JSON serializable object.
+   * @returns {{itemCount: number}} The state of this block, ie the item count.
+   */
+  saveExtraState: function () {
+    return {
+      itemCount: this.itemCount_,
+    };
+  },
+
+  /**
+   * Applies the given state to this block.
+   * @param {*} state The state to apply to this block, ie the item count.
+   */
+  loadExtraState: function (state) {
+    this.updateShape_(state['itemCount']);
+  },
+
+  /**
+   * Adds inputs to the block until the block reaches the target input count.
+   * @param {number} targetCount The number of inputs the block should have.
+   * @this {GoogleBlockly.Block}
+   * @private
+   */
+  updateShape_: function (targetCount) {
+    targetCount = Math.max(targetCount, 1);
+    while (this.itemCount_ < targetCount) {
+      this.addPart_();
+    }
+    while (this.itemCount_ > targetCount) {
+      this.removePart_();
+    }
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the plus image. Adds an input to the block and updates the
+   * state of the minus.
+   * @this {GoogleBlockly.Block}
+   */
+  plus: function () {
+    this.addPart_();
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the minus image. Removes the input at the end of the block and
+   * updates the state of the minus.
+   * @this {GoogleBlockly.Block}
+   */
+  minus: function () {
+    if (this.itemCount_ <= MINIMUM_INPUTS) {
+      return;
+    }
+    this.removePart_();
+    this.updateMinus_();
+  },
+
+  /**
+   * Adds an input to the end of the block. If the block currently has no
+   * inputs it updates the top 'EMPTY' input to receive a block.
+   * @this {GoogleBlockly.Block}
+   * @private
+   */
+  addPart_: function () {
+    if (this.itemCount_ === 0) {
+      if (this.getInput('EMPTY')) {
+        this.removeInput('EMPTY');
+      }
+      this.topInput_ = this.appendValueInput('ADD' + this.itemCount_)
+        // Use @cdo/locale to get localized 'join' string
+        .appendField(msg.joinText());
+    } else {
+      this.appendValueInput('ADD' + this.itemCount_);
+    }
+    if (this.itemCount_ === MINIMUM_INPUTS - 1) {
+      this.iconInput_ = this.getIconInput();
+      this.iconInput_.insertFieldAt(0, createPlusField(), 'PLUS');
+    }
+    // Because item inputs are 0-index we decrement first, increment last.
+    this.itemCount_++;
+  },
+
+  /**
+   * Removes an input from the end of the block. If we are removing the last
+   * input this updates the block to have an 'EMPTY' top input.
+   * @this {GoogleBlockly.Block}
+   * @private
+   */
+  removePart_: function () {
+    this.itemCount_--;
+    this.removeInput('ADD' + this.itemCount_);
+    if (this.itemCount_ === 0) {
```

Comment:
why can we go down to 0 items when the minimum inputs is 2?

Reply:
We can’t, as it’s written, but there was no minimum in the original plugin. I thought it better to leave that functionality (but make in accessible) for now in case we wanted to support the curriculum in setting a minimum per-level or per-block. I also had the thought that the Blockly team might want to add that level of customization to the plugin if they update it. Happy to strip it down if those are bad/confusing reasons!